util.c

```diff
@@ -306,6 +330,26 @@ vg_encode_privkey(const EC_KEY *pkey, int addrtype, char *result)
 	vg_b58_encode_check(eckey_buf, 33, result);
 }
 
+void
+vg_encode_privkey_compressed(const EC_KEY *pkey, int addrtype, char *result)
+{
+	unsigned char eckey_buf[128];
+	const BIGNUM *bn;
+	int nbytes;
+
+	bn = EC_KEY_get0_private_key(pkey);
+
+	eckey_buf[0] = addrtype;
+	nbytes = BN_num_bytes(bn);
+	assert(nbytes <= 32);
+	if (nbytes < 32)
+		memset(eckey_buf + 1, 0, 32 - nbytes);
+	BN_bn2bin(bn, &eckey_buf[33 - nbytes]);
+	eckey_buf[nbytes+1] = 1;
```

Comment:
I know this is an older PR, but I don't think this is correct. If bn does not have 32 bytes filled, this line will end up setting the 1 in the wrong place.

An example case where this happens:

```
// privkey that represents the hex below
puts("KwDso3DYHrJGKGFXmmEWwrj4mkxhp9Vj7ZwYUzRxykqMnccPjLRn");
BIGNUM *bn = BN_new();
BN_hex2bn(&bn, "0015d21a7852d7a408af01c9408895108842c373e88cbff1c9abf27871144938");
EC_KEY *pkey = EC_KEY_new_by_curve_name(NID_secp256k1);
EC_KEY_set_private_key(pkey, bn);
char privkey_buf[VG_PROTKEY_MAX_B58];
vg_encode_privkey_compressed(pkey, 128, privkey_buf);
puts(privkey_buf);
```

Running this prints the following:

```
KwDso3DYHrJGKGFXmmEWwrj4mkxhp9Vj7ZwYUzRxykqMnccPjLRn
KwDso3DYHrJGKGFXmmEWwrj4mkxhp9Vj7ZwYUzRxykqMKHREknvq
```

While this example is outside of vanitygen, I have also been able to reproduce the behavior using it (vanitygen prints a compressed privkey that does not correlate to the matched address).

I believe the solution is setting this index to always be 33:

```
eckey_buf[33] = 1
```

@WyseNynja, this may be relevant to your fork as well.